psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1791,597 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+            disps[atom][point][1] = r_nuc[1];
+            disps[atom][point][2] = r_nuc[2];
+        }
+    }
+
+    // => Setup Proatom Basis Functions <= //
+
+    // mA is the number of shells in an isolated atom
+    std::vector<int> mA(num_atoms);
+    for (int atom = 0; atom < num_atoms; atom++) {
+        int atomic_num = (int) mol->Z(atom);
+        if (atomic_num <= 2) {
+            mA[atom] = 1;
+        } else if (atomic_num <= 10) {
+            mA[atom] = 2;
+        } else if (atomic_num <= 18) {
+            mA[atom] = 3;
+        } else if (atomic_num <= 36) {
+            mA[atom] = 4;
+        } else {
+            // TODO: good guesses for past second row
+            throw PsiException("Atomic Number " + std::to_string(atomic_num) + " unsupported by MBIS", __FILE__, __LINE__);
+        }
+    }
+    
+    // Atomic shell populations (N) and widths (S) (up to 4 shells per atom), from equations 18 and 19 in Verstraelen et al.
+    std::vector<std::vector<double>> Nai(num_atoms, std::vector<double>(4, 0.0));
+    std::vector<std::vector<double>> Sai(num_atoms, std::vector<double>(4, 0.0));
+
+    // Next iteration populations and widths
+    std::vector<std::vector<double>> Nai_next(num_atoms, std::vector<double>(4, 0.0));
+    std::vector<std::vector<double>> Sai_next(num_atoms, std::vector<double>(4, 0.0));
+
+    // Population and width guesses, from get_mbis_params function
+    for (int atom = 0; atom < num_atoms; atom++) {
+        int atomic_num = (int) mol->Z(atom);        
+
+        for (int m = 0; m < mA[atom]; m++) {
+            std::tie(Nai[atom][m], Sai[atom][m]) = get_mbis_params(atomic_num, m);
+            Sai[atom][m] = 1.0/Sai[atom][m];
+            
+            if (print_output && debug >= 1) {
+                outfile->Printf("  INITIAL ATOM %d, SHELL %d, POP %8.5f, WIDTH %8.5f\n", atom+1, m+1, Nai[atom][m], Sai[atom][m]);
+            }
+            
+            Nai_next[atom][m] = Nai[atom][m];
+            Sai_next[atom][m] = Sai[atom][m];
+        }
+    }
+
+    // Promolecular and proatomic densities
+    std::vector<double> rho_0_points(total_points, 0.0);
+    std::vector<std::vector<double>> rho_a_0_points(num_atoms, std::vector<double>(total_points, 0.0));
+
+    // => Main Stockholder Loop <= //
+
+    int iter = 1;
+    bool is_converged = false;
+    double delta_rho_max_0;
+
+    if (print_output && debug >= 1) outfile->Printf("                     Delta D\n");
+    while (iter < max_iter) {
+
+        // Calculate proatom and promolecule density at all points
+#pragma omp parallel for
+        for (int point = 0; point < total_points; point++) {
+            rho_0_points[point] = 0.0;
+            for (int atom = 0; atom < num_atoms; atom++) {
+                rho_a_0_points[atom][point] = 0.0;
+                for (int m = 0; m < mA[atom]; m++) {
+                    rho_a_0_points[atom][point] += rho_ai_0(Nai[atom][m], Sai[atom][m], distances[atom][point]);
```

Comment:
The array indices are in the wrong order, so the memory access here is very inefficient.